### PR TITLE
support for multiple CVMFS repositories in `eessi_container.sh`

### LINKED ISSUE #616: [dev.eessi.io] enable `eessi_container.sh` to mount several CVMFS repos with individual access mode
This is a prerequisite for `dev.eessi.io`.

Currently, `eessi_container.sh` mounts two CVMFS repositories:
- /cvmfs/cvmfs-config.cern.ch
- `SOME` software repository provided via `--repository` [default: /cvmfs/software.eessi.io]

If we want to support building for other repositories particularly on top of `SOME`, the script needs to be changed
- to allow and mount multiple CVMFS repositories instead of just one
- to enable access modes to be individually set for each repository (`rw` - for read & write, `ro` - for read-only)

For each CVMFS repository to be mounted, the script should manage sets of parameters and temporary directories. It should also store these settings somewhere such that we can easily resume with the same configuration.

The format for the value of `--repository` could be changed by supporting a suffix such as `,access=ACCESS_MODE` to specify the access mode for a repository. May want to add another suffix to provide repository-specific `host_injections` (`,host_injections=HOST_PATH:CONTAINER_PATH`).

The script sources other scripts of `EESSI/software-layer`, e.g., `init/eessi_defaults`. The need and way to source these scripts may need to be revisited (are they needed, do they need to be changed too, etc).

### PR DESCRIPTION
**Pull request by trz42**
Support for multiple CVMFS repositories is needed for `dev.eessi.io`.

Ready for review. Example test:
```bash
# ./eessi_container.sh --repository software.eessi.io --repository dev.eessi.io,access=rw
Using /tmp/eessi.jLuBerqoFu as tmp directory (to resume session add '--resume /tmp/eessi.jLuBerqoFu').
Pulling container image from docker://ghcr.io/eessi/build-node:debian11 to /tmp/eessi.jLuBerqoFu/ghcr.io_eessi_build_node_debian11.sif
Launching container with command (next line):
singularity -q shell  --fusemount container:cvmfs2 cvmfs-config.cern.ch /cvmfs/cvmfs-config.cern.ch --fusemount container:cvmfs2 software.eessi.io /cvmfs/software.eessi.io --fusemount container:cvmfs2 dev.eessi.io /cvmfs_ro/dev.eessi.io --fusemount container:fuse-overlayfs -o lowerdir=/cvmfs_ro/dev.eessi.io -o upperdir=/tmp/dev.eessi.io/overlay-upper -o workdir=/tmp/dev.eessi.io/overlay-work /cvmfs/dev.eessi.io /tmp/eessi.jLuBerqoFu/ghcr.io_eessi_build_node_debian11.sif
INFO:    Environment variable SINGULARITY_BIND is set, but APPTAINER_BIND is preferred
INFO:    Environment variable SINGULARITY_HOME is set, but APPTAINER_HOME is preferred
CernVM-FS: pre-mounted on file descriptor 3
CernVM-FS: pre-mounted on file descriptor 3
CernVM-FS: pre-mounted on file descriptor 3
Apptainer> CernVM-FS: loading Fuse module... done
CernVM-FS: loading Fuse module... CernVM-FS: loading Fuse module... done
done

Apptainer> touch /cvmfs/dev.eessi.io/foo
Apptainer> touch /cvmfs/software.eessi.io/foo
touch: cannot touch '/cvmfs/software.eessi.io/foo': Permission denied
```

Partially fixes #616 (doesn't provide repository-specific bind mounts for `host_injections`).

### CODE REVIEW

**Comment by Neves-P** [comment]
Tried the container with @trz42's instructions from slack 
``` shell
# Step 1: checkout repo (and PR)
git clone https://github.com/EESSI/software-layer
cd software-layer
git fetch origin pull/618/head:PR618
git checkout PR618
# Step 2: launch interactive job, for example
srun --time=60 --pty bash
# Step 3: launch container with read access to software.eessi.io and write access to dev.eessi.io
./eessi_container.sh -r software.eessi.io -r dev.eessi.io,access=rw
# Step 4: init EESSI and build environment
source /cvmfs/software.eessi.io/versions/2023.06/init/bash
export EESSI_PROJECT_INSTALL=/cvmfs/dev.eessi.io/dev_test/versions/2023.06/software/linux/${EESSI_SOFTWARE_SUBDIR}
module load EESSI-extend/2023.06-easybuild
# Step 5: build some software (e.g., recently merged imbalanced-learn-0.12.3-gfbf-2023a.eb)
eb --robot --from-pr 20848 imbalanced-learn-0.12.3-gfbf-2023a.eb
# Step 6: show, load and test module
module show imbalanced-learn/0.12.3-gfbf-2023a
module load imbalanced-learn/0.12.3-gfbf-2023a
python -c "import imblearn"
```

Works out of the box! The only thing extra I had to do was: `mkdir -p /cvmfs/dev.eessi.io/dev_test/versions/2023.06/software/linux/${EESSI_SOFTWARE_SUBDIR}`. 

Otherwise I get: 
```
{EESSI 2023.06} Apptainer> module load EESSI-extend/2023.06-easybuild
-- Using /tmp/$USER as a temporary working directory for installations, you can override this by setting the
environment variable WORKING_DIR and reloading the module (e.g., /dev/shm is a common option)
Lmod has detected the following error:  The location of EESSI_PROJECT_INSTALL
(/cvmfs/dev.eessi.io/dev_test/versions/2023.06/software/linux/aarch64/neoverse_v1) does not exist or is not a folder
While processing the following module(s):
    Module fullname                 Module Filename
    ---------------                 ---------------
    EESSI-extend/2023.06-easybuild  /cvmfs/software.eessi.io/versions/2023.06/software/linux/aarch64/neoverse_v1/modules/all/EESSI-extend/2023.06-easybuild.lua
```

Once I first created the directory then I was able to load `EESSI-extend` in such a way that newly installed modules would land in `dev.eessi.io`, which is the intended behaviour.

I might be missing a step, but if not perhaps there is a way to take care of creating the necessary subdirectories automatically (i.e., `cvmfs/somerepo/right_path/${EESSI_SOFTWARE_SUBDIR}`?

**Review by Neves-P — DISMISSED** [review]
Quoting my message from slack:
> I can't see anything wrong in the code, but it might be a good idea to have someone who is more familiar with setting up CVMFS repositories than I am to also take a look since some changes may have implications that I'm not aware of. Beyond that, looks good to me, and works!

File: eessi_container.sh [hunk 1]
```diff
@@ -428,6 +510,7 @@ fi
 [[ ${VERBOSE} -eq 1 ]] && echo "CONTAINER=${CONTAINER}"
 
 # set env vars and create directories for CernVM-FS
+# TODO need to use separate values for separate repos?
```
Comment (bedroge) [inline eessi_container.sh:513]: I don't think that's really necessary, it's fine to have just one cache directory.

File: eessi_container.sh [hunk 2]
```diff
@@ -89,8 +90,11 @@ display_help() {
   echo "                            MODE==install for a CUDA installation, MODE==run to"
   echo "                            attach a GPU, MODE==all for both [default: false]"
   echo "  -r | --repository CFG   - configuration file or identifier defining the"
-  echo "                            repository to use [default: EESSI via"
-  echo "                            default container, see --container]"
+  echo "                            repository to use; can be given multiple times;"
+  echo "                            CFG may include a suffix ';access={ro,rw}' to"
```
Comment (bedroge) [inline eessi_container.sh:94]:
```suggestion
  echo "                            CFG may include a suffix ',access={ro,rw}' to"
```

File: eessi_container.sh (hunk 1 above)
Comment (bedroge) [reply eessi_container.sh:513]:
```suggestion
```

**Review by bedroge — DISMISSED** [review]
Looks good to me. Tested it with various scenarios, and it worked like a charm!

**Review by Neves-P — APPROVED** [review]
LGTM!

**Comment by bedroge** [comment]
The "tests for eessi_container.sh script / eessi_container_script (listrepos_custom)" CI was actually already broken: it was configuring a "EESSI/20AB.CD" and "EESSI/20HT.TP", but then running a grep for ""[EESSI/2023.02]". Not only is that version not available, also the regex itself is wrong: the square brackets should have been escaped, otherwise they will just try to match any of those characters. Because of the wrong regex, the check still passed with this wrong EESSI version.

I've fixed it in https://github.com/EESSI/software-layer/pull/618/commits/aa9895e44ac1afd5a44524d004161f3804449a90. Also, the other check now doesn't look for "EESSI" anymore (which used to be the default repo name), but instead for "software.eessi.io". This was needed because of the changes in this PR.

**Review by Neves-P — APPROVED** [review]
Regex is fixed and the CI passed without issues 🎉

File: eessi_container.sh [hunk 3]
```diff
@@ -489,91 +572,111 @@ fi
...
+            tar xf ${config_bundle_path} -C ${EESSI_TMPDIR}/repos_cfg
+        fi
 
-  for src in "${!cfg_file_map[@]}"
-  do
-    target=${cfg_file_map[${src}]}
-    BIND_PATHS="${BIND_PATHS},${EESSI_TMPDIR}/repos_cfg/${src}:${target}"
-  done
-  export EESSI_VERSION_OVERRIDE=${repo_version}
-  export EESSI_CVMFS_REPO_OVERRIDE="/cvmfs/${repo_name}"
```
Comment (boegel) [inline eessi_container.sh:569]: Seems like this was important, it shouldn't have been removed.

I'm now seeing errors like this in the build job for https://github.com/EESSI/software-layer/pull/632:
```
ERROR: software.eessi.io/versions/2023.06/compat/linux/aarch64 does not exist!
```

Note the missing `/cvmfs`...

Comment (boegel) [reply eessi_container.sh:569]: Or maybe we're missing a fix in `bot/build.sh`, which now has this:
```shell
export EESSI_CVMFS_REPO_OVERRIDE=$(cfg_get_value "repository" "repo_name")
```

I guess this should be:
```shell
export EESSI_CVMFS_REPO_OVERRIDE=/cvmfs/$(cfg_get_value "repository" "repo_name")
```

Comment (bedroge) [reply eessi_container.sh:569]: Ah, yes, good catch, looks like that needs to be changed (and I think it makes more sense that these things are (only) set there and not in the container script).

Comment (boegel) [reply eessi_container.sh:569]: see #633

**Pull request merged**